File: Cpp/src/repositories/UserRepository.cc
#include "UserRepository.h"
#include "../db/Database.h"
#include "../db/StmtGuard.h"
#include <iostream>

namespace repositories {

    int UserRepository::create(const models::User& user) {
        sqlite3* db = db::Database::getInstance().getDb();
        sqlite3_stmt* stmt;
        const char* sql = "INSERT INTO users (name, email) VALUES (?, ?);";

        if (sqlite3_prepare_v2(db, sql, -1, &stmt, 0) != SQLITE_OK) {
             throw std::runtime_error("Failed to prepare statement");
        }
        db::StmtGuard guard(stmt);

        sqlite3_bind_text(stmt, 1, user.name.c_str(), -1, SQLITE_STATIC);
        sqlite3_bind_text(stmt, 2, user.email.c_str(), -1, SQLITE_STATIC);

        if (sqlite3_step(stmt) != SQLITE_DONE) {
            throw std::runtime_error("Failed to execute statement");
        }

        return (int)sqlite3_last_insert_rowid(db);
    }

    std::optional<models::User> UserRepository::findById(int id) {
        sqlite3* db = db::Database::getInstance().getDb();
        sqlite3_stmt* stmt;
        const char* sql = "SELECT id, name, email FROM users WHERE id = ?;";

        if (sqlite3_prepare_v2(db, sql, -1, &stmt, 0) != SQLITE_OK) {
             throw std::runtime_error("Failed to prepare statement");
        }
        db::StmtGuard guard(stmt);

        sqlite3_bind_int(stmt, 1, id);

        if (sqlite3_step(stmt) == SQLITE_ROW) {
            models::User user;
            user.id = sqlite3_column_int(stmt, 0);
            user.name = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 1));
            user.email = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 2));
            return user;
        }

        return std::nullopt;
    }

    std::vector<models::User> UserRepository::findAll() {
        std::vector<models::User> users;
        sqlite3* db = db::Database::getInstance().getDb();
        sqlite3_stmt* stmt;
        const char* sql = "SELECT id, name, email FROM users;";

        if (sqlite3_prepare_v2(db, sql, -1, &stmt, 0) != SQLITE_OK) {
             throw std::runtime_error("Failed to prepare statement");
        }
        db::StmtGuard guard(stmt);

        while (sqlite3_step(stmt) == SQLITE_ROW) {
            models::User user;
            user.id = sqlite3_column_int(stmt, 0);
            user.name = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 1));
            user.email = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 2));
            users.push_back(user);
        }

        return users;
    }

    void UserRepository::update(const models::User& user) {
        sqlite3* db = db::Database::getInstance().getDb();

        std::string sql = "UPDATE users SET ";
        bool first = true;
        bool hasUpdate = false;

        if (!user.name.empty()) {
            sql += "name = ?";
            first = false;
            hasUpdate = true;
        }
        if (!user.email.empty()) {
            if (!first) sql += ", ";
            sql += "email = ?";
            hasUpdate = true;
        }

        if (!hasUpdate) return; // Nothing to update

        sql += " WHERE id = ?;";

        sqlite3_stmt* stmt;
        if (sqlite3_prepare_v2(db, sql.c_str(), -1, &stmt, 0) != SQLITE_OK) {
             throw std::runtime_error("Failed to prepare statement");
        }
        db::StmtGuard guard(stmt);

        int index = 1;
        if (!user.name.empty()) {
            sqlite3_bind_text(stmt, index++, user.name.c_str(), -1, SQLITE_STATIC);
        }
        if (!user.email.empty()) {
            sqlite3_bind_text(stmt, index++, user.email.c_str(), -1, SQLITE_STATIC);
        }
        sqlite3_bind_int(stmt, index, user.id);

        if (sqlite3_step(stmt) != SQLITE_DONE) {
            throw std::runtime_error("Failed to execute statement");
        }
    }

    void UserRepository::deleteById(int id) {
        sqlite3* db = db::Database::getInstance().getDb();
        sqlite3_stmt* stmt;
        const char* sql = "DELETE FROM users WHERE id = ?;";

        if (sqlite3_prepare_v2(db, sql, -1, &stmt, 0) != SQLITE_OK) {
             throw std::runtime_error("Failed to prepare statement");
        }
        db::StmtGuard guard(stmt);

        sqlite3_bind_int(stmt, 1, id);

        if (sqlite3_step(stmt) != SQLITE_DONE) {
            throw std::runtime_error("Failed to execute statement");
        }
    }

    bool UserRepository::exists(int id) {
        return findById(id).has_value();
    }
}

File: LuaOpenResty/app/repositories/user_repository.lua
local DB = require("db.db")
local User = require("app.models.user")

local _M = {}

function _M.create(username, email)
    local db = DB.connect()
    local res, err = db:execute("INSERT INTO users (username, email) VALUES (?, ?)", {username, email})
    db:close()

    if not res then return nil, err end
    return User.new(res.last_insert_id, username, email), nil
end

function _M.find_all()
    local db = DB.connect()
    local rows, err = db:query("SELECT id, username, email FROM users")
    db:close()

    if not rows then return {}, err end

    local users = {}
    for _, row in ipairs(rows) do
        table.insert(users, User.new(row.id, row.username, row.email))
    end
    return users
end

function _M.find_by_id(id)
    local db = DB.connect()
    local rows, err = db:query("SELECT id, username, email FROM users WHERE id = ?", {id})
    db:close()

    if not rows or #rows == 0 then return nil, err end
    local row = rows[1]
    return User.new(row.id, row.username, row.email)
end

function _M.update(id, username)
    local db = DB.connect()
    local res, err = db:execute("UPDATE users SET username = ? WHERE id = ?", {username, id})
    db:close()

    if not res then return nil, err end
    if res.changes == 0 then return nil, "not found" end
    return true
end

function _M.delete(id)
    local db = DB.connect()
    local res, err = db:execute("DELETE FROM users WHERE id = ?", {id})
    db:close()

    if not res then return nil, err end
    return true
end

return _M
